### PR TITLE
Zhanfu/csr Framework for the CSR structure

### PR DESCRIPTION
**Pull request by peter-rich**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 1]
```diff
@@ -12,77 +12,61 @@ namespace graph {
...
+   unsigned int e_count = iterDE->size();
 
-   ResultBuilder builder;
-   builder.value(iter->valuePtr());
+   unsigned int *offset = new unsigned int[v_count + 2];
+   unsigned int *neighbors = new unsigned int[e_count * 2];
+   unsigned int *labels = new unsigned int[l_count];
+   // load data vertices id and tags
+   while (iterDV->valid()) {
+     const auto v = iterDV->getColumn(nebula::kVid);  // check if v is a vertex
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:33]: what are v and v1. Please use meaningful names.

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 2]
```diff
@@ -12,77 +12,61 @@ namespace graph {
...
 
-   auto iter = ectx_->getResult(subgraph->inputVar()).iter();
-   auto gnSize = iter->size();
+   unsigned int v_count = iterDV->size();
+   unsigned int l_count = iterDV->size();
+   unsigned int e_count = iterDE->size();
 
-   ResultBuilder builder;
-   builder.value(iter->valuePtr());
+   unsigned int *offset = new unsigned int[v_count + 2];
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:28]: (1) Please explain via comments why the size is +2
(2) Please use vector. Don't use new.

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 3]
```diff
@@ -12,77 +12,61 @@ namespace graph {
...
-     }
+
+   // load data edges
+   offset[0] = 0;
+   iterDE = ectx_->getResult(isomor->getdScanEOut()).iter();
+   while (iterDE->valid()) {
+     unsigned int src = iterDE->getEdgeProp("*", kSrc).getInt();
+     unsigned int dst = iterDE->getEdgeProp("*", kDst).getInt();
+
+     neighbors[offset[src + 1]] = dst;
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:57]: Do not understand this. Please explain via comment.

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 4]
```diff
@@ -12,77 +12,61 @@ namespace graph {
...
-   auto iter = ectx_->getResult(subgraph->inputVar()).iter();
-   auto gnSize = iter->size();
+   unsigned int v_count = iterDV->size();
+   unsigned int l_count = iterDV->size();
+   unsigned int e_count = iterDE->size();
 
-   ResultBuilder builder;
-   builder.value(iter->valuePtr());
+   unsigned int *offset = new unsigned int[v_count + 2];
+   unsigned int *neighbors = new unsigned int[e_count * 2];
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:29]: Remember that for every edge A-B, we actually store 2 copies, A->B and B->A。 You need to take this into consideration. They can be distinguished by edge type signs. Positive and Negative represent them respectively. Please think about it.

File: src/graph/executor/algo/IsomorExecutor.cpp (hunk 1 above)
Comment (peter-rich) [reply src/graph/executor/algo/IsomorExecutor.cpp:33]: I change it to the new version in the code, v is the content of the vertex id while v2 is the content of the label id.

File: src/graph/executor/algo/IsomorExecutor.cpp (hunk 2 above)
Comment (peter-rich) [reply src/graph/executor/algo/IsomorExecutor.cpp:28]: (1) as the offset is recording the starting position of the node 0 which suppose to be 0, and offset[1] should be the end of position node 0 and start of position node 1. The size of the offset should be at least v_count + 1;
(2) Ok

File: src/graph/executor/algo/IsomorExecutor.cpp (hunk 4 above)
Comment (peter-rich) [reply src/graph/executor/algo/IsomorExecutor.cpp:29]: Ok

File: src/graph/executor/algo/IsomorExecutor.cpp (hunk 3 above)
Comment (peter-rich) [reply src/graph/executor/algo/IsomorExecutor.cpp:57]: (1) src and dst is the source and destination information.
(2) offset[src + 1] will be the position of the array.

Comment (peter-rich) [reply src/graph/executor/algo/IsomorExecutor.cpp:57]: I can change it and comment it in the new version

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 5]
```diff
@@ -1,51 +1,46 @@
...
-     const auto v = iterDV->getColumn(nebula::kVid);  // check if v is a vertex
-     auto v_id = v.getInt();
-     const auto v2 = iterDV->getColumn(1);  // get label
-     auto l_id = v2.getInt();
+     const auto vertex = iterDV->getColumn(nebula::kVid);  // check if v is a vertex
+     auto v_id = vertex.getInt();
+     const auto label = iterDV->getColumn(1);  // get label by index
+     auto l_id = label.getInt();
      // unsigned int v_id = (unsigned int)v.getInt(0);
      labels[v_id] = l_id;  // Tag Id
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:31]: Can vID used as the index of the `labels` array? There is no restriction on what vID will look like, e.g., may be from 1000 to 100000.

Comment (wenhaocs) [reply src/graph/executor/algo/IsomorExecutor.cpp:31]: Therefore, it's better to use map here.

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 6]
```diff
@@ -1,88 +1,152 @@
...
+  }
+  // load edges degree
+  while (iterDE->valid()) {
+    auto s = iterDE->getEdgeProp("*", kSrc);
+    unsigned int src = s.getInt();
+    offset[src]++;
+    iterDE->next();
+  }
+  for (unsigned int i = 0; i < v_count; i++) {
+    offset[i + 1] = offset[i];
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:53]: Is this wrong?

File: src/graph/executor/CMakeLists.txt [hunk 7]
```diff
@@ -90,4 +93,4 @@ nebula_add_library(
     mutate/UpdateExecutor.cpp
 )
 
-nebula_add_subdirectory(test)
```
Comment (wenhaocs) [inline src/graph/executor/CMakeLists.txt:93]: why is this line deleted

File: src/graph/executor/algo/IsomorExecutor.cpp [hunk 8]
```diff
@@ -50,13 +66,13 @@ folly::Future<Status> IsomorExecutor::execute() {
...
     unsigned int src = s.getInt();
-    offset[src]++;
+    degree[src]++;
     iterDE->next();
   }
 
   // caldulate the start position of each vertex in the neighborhood array
   for (unsigned int i = 0; i < v_count; i++) {
-    offset[i + 1] += offset[i];
+    offset[i + 1] += degree[i] + offset[i];
```
Comment (wenhaocs) [inline src/graph/executor/algo/IsomorExecutor.cpp:75]: +=？
